### DISCUSSION TITLE
Raise exception from timer interrupt routine Micropython RP2040

### DISCUSSION THREAD

**Discussion by John1900** [micropython/micropython#9229] [Using MicroPython]
Is it possible to raise an exception from a timer interrupt?

Having tried variations on this code there is likely something I just don't get!   Thanks

        timer = Timer(-1)
        timer.init(period=1000, mode=Timer.ONE_SHOT, callback=lambda t:raise Exception("Error"))

**Comment by jimmo** [comment]
Just to clarify, what do you expect to happen with the raised exception? Interrupt the main thread?

~That's what will happen... an exception will be scheduled to be raised when the VM resumes operating after the IRQ finishes.~ (Edit: see later post. Soft interrupts and scheduled tasks have their own catch-all exception handlers)

**Reply by John1900** [reply]
Yes, raise the exception.  The timeout will occur because a series of operations using servo motors will not have completed in a reasonable time (indicated by a switch state).   The timer should be easier than checking individually the time taken for each operation.  Without checking the system will just hang when a server jams or something else goes wrong.

**Comment by robert-hh** [comment]
Just for interest I tried that. For some reason I could not put the raise into the lambda function. But in  separate small function it worked.
```
from machine import Timer
import time

def r(t):
    raise Exception("Error")

timer = Timer(-1, period=3000, mode=Timer.ONE_SHOT, callback=r)

while 1:
    try:
        print("Wait for the exception")
        time.sleep(1)
    except Exception as e:
        print("Exception", e)

```

**Reply by John1900** [reply]
Thanks but that shows another problem.  The "except" code never runs and no way to exit the while loop:

```
a = True
while a:
    try:
        print("Wait for the exception")
        time.sleep(1)
    except Exception as e:
        print("Exception here", e)
        a = False
        break
```

**Reply by John1900** [reply]
```py
from machine import Timer
import time
def r(t):
     raise Exception("Error")

timer = Timer(-1, period=3000, mode=Timer.ONE_SHOT, callback=r)

a = True
while a:
    try:
        print("Wait for the exception")
        time.sleep(1)
    except Exception as e:
        print("Exception here", e)
        a = False
        break
```

Sorry about the formatting!  (Edit by @jimmo to fix formatting)

**Reply by robert-hh** [reply]
About formatting:
you have to enclose code into three backticks. An easy way of doing that is to highlight the pasted code and push Ctrl-E.
 
But interestingly, this does not work either. The `break` does not.

**Reply by jimmo** [reply]
@John1900 Please see https://github.com/orgs/micropython/discussions/9111 for more info about code formatting.

**Comment by robert-hh** [comment]
Here, the except code runs, as the print statement is executed. But the loop does not stop, since a is not modified. Putting all into a class does not help either. Hmm...

**Reply by John1900** [reply]
I also thought the except code ran so try changing to   print("Exception here", e)    I only noticed cause it prints with a colon which was not in your print.  Yes, Hmm...

**Comment by jimmo** [comment]
Sorry! I said the wrong thing in my first reply.

The timer runs its callback via the scheduler (it's a soft interrupt in the context of https://docs.micropython.org/en/latest/reference/isr_rules.html ). Scheduled tasks have a built-in catch-all exception handler which just prints the exception out directly.

If you want to signal from an interrupt handler, then you should do so with a global variable.

```py
from machine import Timer
import time

timeout = False

def r(t):
    global timeout
    timeout = True

timer = Timer(-1, period=3000, mode=Timer.ONE_SHOT, callback=r)

while True:
    if timeout:
        print("Timeout")
        break
    # otherwise keep working
```

In general though it's likely easier to implement timeouts by measuring time taken..

```py
import time

t_start = time.ticks_ms()
while True:
    if time.ticks_diff(time.ticks_ms(), t_start) > 3000:
        print("Timeout")
        break
    # otherwise keep working
```

**Reply by John1900** [reply]
Thanks, works great, just what I needed.

I agree ticks_diff is much easier but I have several places with;  "while something is happening/sleep".  Its much easier to have one timer for a group of operations which I can cancel on normal completion - rather than add time code to each operation.
